### PR TITLE
Fix FYSETC S6/Spider boot issue

### PR DESCRIPTION
**Pull request by GerogeFu**
Fix the issue reported here: https://github.com/MarlinFirmware/Marlin/issues/22345

### CODE REVIEW

**Comment by thisiskeithb** [comment]
Based on @GMagician's comment in https://github.com/MarlinFirmware/Marlin/issues/22345#issuecomment-885135744, this patch may not work. Can anyone else confirm if this fixes the no boot issue on the S6 & Spider?

**Comment by GMagician** [comment]
What I see, loading elf file, is that address seems wrong. Entry point is 45 20 01 08 and loading elf file with st cube it's loaded at `0x08010000`.

**Comment by tpruvot** [comment]
beware, its 0x0801.... not 0x8001...

**Comment by GMagician** [comment]
That's right but S6 env (I compiled both) has the same entry point, so it's really unusual

**Comment by GMagician** [comment]
Oh wait, maybe my fault. I changed wrong ldscrit file!

**Comment by rsdcampos** [comment]
Hi guys, after a lot of attempts, finally I could make it to work with new bootloader, but the eeprom still don't saves the data when I sent M500.
```
echo:Error writing to EEPROM!
echo:Settings Stored (7 bytes; crc 53868)
Error:EEPROM datasize error.
```
Any with same issue and know how fix it?
The precompiled one works, but when I compile have this issue in both flash shift adress.

---
What I did:
First of all, I used the last stable Marlin (2.0.9.1);
In `buildroot\share\PlatformIO\variants`, I duplicated the `MARLIN_FYSETC_S6` folder and renamed it to `MARLIN_FYSETC_S6_8000`;
In this folder I changed in `ldscript.ld` the flash address to `0x08008000`.
In `buildroot\share\PlatformIO\boards\marlin_fysetc_s6_8000.json` I changed the variant from `MARLIN_FYSETC_S6` to `MARLIN_FYSETC_S6_8000`.
I don't know if there are a better way to do it, but it was the way I made it work.

**Comment by rsdcampos** [comment]
What I did:
First of all, I used the Lest stable Marlin (2.0.9.1);
In buildroot\share\PlatformIO\variants, I duplicate the MARLIN_FYSETC_S6 folder and renamed it to MARLIN_FYSETC_S6_8000;
In this folder I changed in ldscript.ld the flash adress to 0x08008000
In buildroot\share\PlatformIO\boards\marlin_fysetc_s6_8000.json I changed the variant from "variant": "MARLIN_FYSETC_S6" to "variant": "MARLIN_FYSETC_S6_8000".
I don't know if there are a better way to do it, but it was the way I did to make it to work.

**Comment by thinkyhead** [comment]
@rsdcampos — A lot has changed since 2.0.9.1 so you ought to be testing with `bugfix-2.0.x` and applying any changes there.

**Comment by rsdcampos** [comment]
Thanks for answer, but the fix is not in bugfix, but I found it. In pins_FYSETC_S6.h the MARLIN_EEPROM_SIZE is defined to 0x0800 (2KB) I changed it to 0x1000 (4KB) and it works now.

**Comment by thinkyhead** [comment]
Ok, I've done my best to bring this up to date with the pattern for other variants. Please test to see that both targets behave properly when paired with their respective bootloaders.

**Comment by rsdcampos** [comment]
@thinkyhead, I'm not a profissional in code and don't know the marlin structure as well as I liked, so I neded to do a trial and error.
now I found a more elegant way to fix it without new variant folder, but see in the changed files here that @GerogeFu found the same, using the LD_FLASH_OFFSET.
@GerogeFu, please, don't forget to change the MARLIN_EEPROM_SIZE from 2K to 4K.
I tested here and works in both variants with it's respectives bootloaders this way.

Thanks everybudy for all.

File: ini/stm32f4.ini [hunk 1]
```diff
@@ -67,11 +67,9 @@ upload_protocol     = dfu
 #
 [env:FYSETC_S6]
 platform                    = ${common_stm32.platform}
-extends                     = stm32_variant
-board                       = marlin_fysetc_s6
-board_build.offset          = 0x10000
-board_upload.offset_address = 0x08010000
-build_flags                 = ${stm32_variant.build_flags} -DHAL_PCD_MODULE_ENABLED
+extends                     = common_stm32
+board                       = fysetc_s6
```
Comment (rsdcampos) [inline ini/stm32f4.ini]: With this changes the board stoped working to me. please try:
```ini
#
# FYSETC S6 (STM32F446RET6 ARM Cortex-M4)
#
[env:FYSETC_S6]
platform                    = ${common_stm32.platform}
extends                     = stm32_variant
board                       = marlin_fysetc_s6
board_build.offset          = 0x10000
board_upload.offset_address = 0x08010000
build_flags                 = ${stm32_variant.build_flags} -DHAL_PCD_MODULE_ENABLED
debug_tool                  = stlink
upload_protocol             = dfu
upload_command              = dfu-util -a 0 -s 0x08010000:leave -D "$SOURCE"

#
# FYSETC S6 new bootloader
#
[env:FYSETC_S6_8000]
platform                    = ${common_stm32.platform}
extends                     = env:FYSETC_S6
board                       = marlin_fysetc_s6
board_build.offset          = 0x8000
board_upload.offset_address = 0x08008000
upload_command              = dfu-util -a 0 -s 0x08008000:leave -D "$SOURCE"
```

Comment (GerogeFu) [reply ini/stm32f4.ini]: I try the branch with my and thinkhead commits, it works in my side. Just keep in mind that you need respective bootloader. If you use FYSETC_S6_8000 env, you need `Bootloader-FYSETC_S6.hex`, `FYSETC_S6` env you need `Bootloader-FYSETC_S6_10000.hex`. You can find the bootloader [here](url). And try this branch https://github.com/GerogeFu/Marlin/tree/boot_issue. And i also try your env, works too.

**Comment by GerogeFu** [comment]
> @thinkyhead, I'm not a profissional in code and don't know the marlin structure as well as I liked, so I neded to do a trial and error.
> now I found a more elegant way to fix it without new variant folder, but see in the changed files here that @GerogeFu found the same, using the LD_FLASH_OFFSET.
> @GerogeFu, please, don't forget to change the MARLIN_EEPROM_SIZE from 2K to 4K.
> I tested here and works in both variants with it's respectives bootloaders this way.
> 
> Thanks everybudy for all.

Yes , i will, eeprom should be 4k. It seems that the commit here https://github.com/MarlinFirmware/Marlin/pull/20340/commits, change the EEPROM SIZE from 4k to 2k, i don't know why it changes. Not matter how, thanks anyway. PR is here: https://github.com/MarlinFirmware/Marlin/pull/22424

**Comment by GMagician** [comment]
If all I did is correct now 8000 boot+env is working but 10000 boot+env doesn't. On latter I see com port but repetier can't connect to it

**Comment by tpruvot** [comment]
maybe board_build.offset = 0x10000 was removed by mistake ?

**Comment by GMagician** [comment]
Nope, adding it doesn't work as well

**Comment by tpruvot** [comment]
You flashed the big vendor bootloader before right ?

**Comment by GMagician** [comment]
yep, of course

**Comment by tpruvot** [comment]
just to be sure, i had to switch often too on my board... and sometimes you forget ;p

**Comment by rsdcampos** [comment]
@GMagician, it is was I tried to explain to @GerogeFu, the changes he made in S6 envs at stm32f4.ini file make it stop working, but just on 0x10000 offset. With the changes I sugested, both bootloaders worked well for me.

**Comment by GMagician** [comment]
@rsdcampos I confirm, restoring .ini on big boot section with previous values (I just keep 'board_upload.offset_address = 0x08010000' removed) big boot env works

**Comment by thinkyhead** [comment]
So, the consensus seems to be that the variant `board = fysetc_s6` provided by the framework is not valid, so we are forced to use our custom variant hack. Is that the case?

**Comment by GMagician** [comment]
> So, the consensus seems to be that the variant board = fysetc_s6 provided by the framework is not valid, so we are forced to use our custom variant hack. Is that the case?

That seems to be

**Comment by thinkyhead** [comment]
Since the `fysetc_s6` board provided by the framework is assumed to be broken, this is now reduced again to minimal changes to accommodate two bootloader sizes. Please confirm that this is now sufficient to cover both cases.

**Comment by GMagician** [comment]
Tested last changes, both envs work!

**Comment by ETE-Design** [comment]
@thinkyhead Can we use Auto Build Marlin, or do we have to do something manual to get the firmware work with 0x8010 bootloader?

**Comment by thisiskeithb** [comment]
> @thinkyhead Can we use Auto Build Marlin, or do we have to do something manual to get the firmware work with 0x8010 bootloader?

You’ll use one of these two environments. ABM will pick them up as well:

https://github.com/MarlinFirmware/Marlin/blob/357fb8d98458883144e29a39e9f4d5c033f803a3/ini/stm32f4.ini#L65-L88

**Comment by ETE-Design** [comment]
@thinkyhead So to understand it correct ABM check by itself if you have New / Old Bootloader on the board and flash at the correct address afterwards?

**Comment by thisiskeithb** [comment]
> @thinkyhead So to understand it correct ABM check by itself if you have New / Old Bootloader on the board and flash at the correct address afterwards?

You'll need to choose the correct environment for your board based on the installed bootloader.

Hop on the [Marlin Firmware](https://discord.gg/n5NJ59y) Discord server if you have more questions.

**Pull request merged**